src/pybind/mgr/dashboard/controllers/home.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import os
+import json
+
+import cherrypy
+from cherrypy.lib.static import serve_file
+
+from . import Controller, UiApiController, BaseController, Proxy, Endpoint
+from .. import mgr, logger
+
+
+LANGUAGES = {f for f in os.listdir(mgr.get_frontend_path()) \
+             if os.path.isdir("{}/{}".format(mgr.get_frontend_path(), f))}
+
+
+def _get_default_language():
+    with open("{}/../package.json".format(mgr.get_frontend_path()), "r") as f:
+        config = json.load(f)
+    return config['config']['locale']
+
+
+DEFAULT_LANGUAGE = _get_default_language()
+
+
+@Controller("/", secure=False)
+class HomeController(BaseController):
+
+    def _parse_accept_language(self):
+        if 'Accept-Language' in cherrypy.request.headers:
+            accept_lang_header = cherrypy.request.headers['Accept-Language']
+        else:
+            return [DEFAULT_LANGUAGE]
+
+        accept_langs = [l.strip() for l in accept_lang_header.split(",")]
+        result = []
+        for accept_lang in accept_langs:
+            if ';' in accept_lang:
+                lang = accept_lang.split(';')[0].strip()
+                ratio = float(accept_lang.split(';')[1].strip()[2:])
+            else:
+                lang = accept_lang.strip()
+                ratio = 1.0
+            result.append((lang, ratio))
+        result.sort(key=lambda l: l[0])
+        result.sort(key=lambda l: l[1], reverse=True)
+        logger.debug("language preference: %s", result)
+        return [l[0] for l in result]
+
+    def _language_dir(self, langs):
+        for lang in langs:
+            if lang in LANGUAGES:
+                logger.debug("found directory for language '%s'", lang)
+                return "{}/{}".format(mgr.get_frontend_path(), lang)
+            for a_lang in LANGUAGES:
+                if a_lang.startswith(lang):
+                    logger.debug("found directory for language '%s'", a_lang)
+                    return "{}/{}".format(mgr.get_frontend_path(), a_lang)
+
+        logger.debug("Languages '%s' not available, falling back to english", langs)
+        return "{}/{}".format(mgr.get_frontend_path(), DEFAULT_LANGUAGE)
```

Comment:
This is now better, but given we are pre-calculating the a mapping of the `LANGUAGES` supported, why not bringing that approach to its logical end? Why not fully pre-calculating the FULL mapping between locales and paths (including the base languages: `en`, `es`, etc.)?

```py
class LocaleDict(dict):                                                                                   
    LOCALE_RE = re.compile(r'(^[:alpha:]{1,8})-([:alpha:]{1,8})')  # RFC 4647
                                                                                                        
    def __init__(self, d):                                                                              
        super(LocaleDict, self).__init__(d)                                                               
                                                                                                        
        for locale, path in d.items():                                                                    
            try:                                                                                        
                lang = self.LOCALE_RE.match(locale)[1]                                               
                if lang not in self:                                                                   
                    self[lang] = path                                                                  
            except TypeError:                                                                           
                pass                                                                                
```
```py
>>> LOCALE_TO_PATH = LocaleDict(LANGUAGE)
>>> LOCALE_TO_PATH['en-US']
'.../dist/en-US'
>>> LOCALE_TO_PATH['en']
'.../dist/en-US'
```
In my opinion we should avoid doing any on-the-fly calculation in the request handler, especially in a request handler that's exposing the filesystem (the only [2 CVEs reported about Cherrypy](https://www.cvedetails.com/vendor/4097/Cherrypy.html) are directory traversal attacks related to the static dir mapping).

It took me a couple of hours of reviewing & testing the previous changeset with different inputs to catch that critical vulnerability. I couldn't find any with this code, but I'm not infallible, so there are still things that concern me:
- We are doing `for/if/else` and string matching logic on a per-request basis, when this could be easily replaced with a simpler, and much less risky, associative mapping: `return LOCALE_TO_PATH[requested_locale]`
- We are not yet properly sanitizing incoming language headers/cookies (`1*8ALPHA-1*8ALPHA` or `[:alpha:]{1,8}-[:alpha:]{1,8}`), as a result of this we are 'doing things' with a string coming from an untrusted source, like injecting it into our Python logger:
  - Allowing an attacker to send arbitrarily large strings (with 100KB `Accept-Language` headers), to cause disk utilization and overall underperformance.
  - Allowing an attacker to make up and send deceiving messages that could confuse an operator (let's say you start seeing your logs plenty of "**`YOUR SYSTEM HAS BEEN HIJACKED`**") and lead them to make wrong decisions (e.g.: **`CEPH: CRITICAL ERROR FOUND: RISK OF DATA LOSS: please type the following command to fix this: <whatever the attacker wants you to do>`**).
- We are joining paths using `str.format`. We should use `os.path.join` or `urllib.urlparse.urljoin` (but never in the request handler).

Reply:
@epuertat should I also restrict the number of language tags that we parse from the `Accept-Language` header? I'm thinking of the case where someone sends an `Accept-Language` header with millions of language tags

Reply:
Yeah, I though that too. I've been checking [how Apache deals with this](https://github.com/apache/httpd/blob/5dc1948b15a17f480c37a4a856c25ad7c383cce7/modules/mappers/mod_negotiation.c#L1431), but it uses a mixed approach (both relying on `q` and on language index) plus fingerprinting, and seems pretty comprehensive but an overkill for our use-case.

Here it comes a few ideas:
- Use an iterator: unfortunately `split()` does not provide an iterator interface, but we can use `re.finditer()`
- Limit the max number of iterations to a small number (10?).
- As the whole expression has lots of optional sub-groups, let's start capturing the comma-separated groups, removing trailing whitespaces (`r'\s*([^,]*)\s*,?\s*'`)
- For the best case (q=1) if Lang build exists return immediately. Otherwise, shall we assume languages are decreasing-q sorted? Or shall we perform
- Use @functools.lrucache to cache the mapping performed. This will save this calculation for the many files the front-end requires to download (>20).

```py
LANGUAGE_TAG_RE = re.compile(r'\s*(?P<language_tag>[^,]*)\s*,?\s*')
LANGUAGE_WEIGHT_RE = re.match(r'^(?P<locale>[a-z]{1,8}(?:-\w{1,8})?|\*)(?:;q=(?P<weight>[01]\.\d{0,3}))?$')

@lru_cache(maxsize=MAX_EXPECTED_PARALLEL_SESSIONS)
def _language_dir(language_header):
  for i, m in enumerate(LANGUAGE_TAG_RE.finditer(language_header)):
    # Safety net
    if i > MAX_LANG_CHECKS:
      return LANGUAGE_TO_PATH[DEFAULT_LANG]

    lang, weight = LANGUAGE_WEIGHT_RE.match(m.group(1)).groups()

    # Best-case: first language is q=1 and exists in LUT
    if lang in LANGUAGE_TO_PATH and (weight is None or float(weight) == 1):
      # Nothing else to do
      return LANGUAGE_TO_PATH[lang]

    # Most browsers specify the languages in decreasing order of q, but HTTP does not requires that :(
    else:
      # return the first matching lang? or save (lang, q) to a sorted array and after iterating, return the best match?
   
```

Reply:
Still thinking on this... After reading an RFC on Content-Negotiation, I think the matter is hardcore stuff, so my suggestion would be to stick to a simpler/safer/working approach, by assuming language tags are sorted by decreasing quality weight (which is true for Mozilla/Webkit browsers), so the first language tag matching a provided build should be returned.

```py
@lru_cache(maxsize=MAX_EXPECTED_PARALLEL_SESSIONS)
def _language_dir(language_header):
  for m in itertools.islice(LANGUAGE_TAG_RE.finditer(language_header), MAX_LANG_CHECKS):
    lang = LANGUAGE_WEIGHT_RE.match(m.group(1)).group(1)
    try:
      return LANGUAGE_TO_PATH[lang]
    except KeyError:
      pass

return LANGUAGE_TO_PATH[DEFAULT_LANG]
```